WooCommerce/Classes/ViewRelated/Products/Edit Product/ProductFormViewController.swift

```diff
@@ -568,7 +568,9 @@ private extension ProductFormViewController {
         defer {
             navigationController?.popViewController(animated: true)
         }
-        guard images != product.images else {
+        let hasChangedData = images != product.images
+        ServiceLocator.analytics.track(.productImageSettingsDoneButtonTapped, withProperties: ["has_changed_data": hasChangedData])
```

Comment:
this is always `false` because `ProductFormViewController` also observes product image changes and updates the product's `images` on changes:

https://github.com/woocommerce/woocommerce-ios/blob/4220aa2efb8ea2764543378b84e53defab03d800/WooCommerce/Classes/ViewRelated/Products/Edit%20Product/ProductFormViewController.swift#L143-L155

so the product's `images` are already updated before the Done CTA is tapped.

to fix this, I think we can pass the value for `has_changed_data` (`productImages  != originalProductImages`) in `ProductImagesViewController`'s completion block:

https://github.com/woocommerce/woocommerce-ios/blob/4220aa2efb8ea2764543378b84e53defab03d800/WooCommerce/Classes/ViewRelated/Products/Media/ProductImagesViewController.swift#L158-L160

Reply:
I just noticed that the method `hasOutstandingChanges()` in `ProductImagesViewController` always returns false. I also see that when I choose a new photo, and I press the navbar button "add," I can't press done. Still, it goes back automatically to the edit product detail (so the Done button in the  `ProductImagesViewController` seems superfluous). Is this the right behavior?

Reply:
I implement it here 26cdcbd4d0edafd3630866acbdbde120c28d92cc but we need another PR where we fix the `hasOutstandingChanges` method, that actually seems not working properly. Related issue: https://github.com/woocommerce/woocommerce-ios/issues/2305

Reply:
ah this is a tricky case, I set some breakpoints and it seems like `productImageStatuses` is updated to include the pending image asynchronously from the uploaded images, and thus we weren't able to get the correct `hasOutstandingChanges` value synchronously from uploading the images.

I think in the cases where we call `commitAndDismiss` in `ProductImagesViewController` from upload completions (`onCameraCaptureCompletion`, `onDeviceMediaLibraryPickerCompletion`, `onWPMediaPickerCompletion`), we can just pass the `hasChangedData` to `true` and skip `hasOutstandingChanges` since these completion blocks are only called if the uploaded images are non-empty

Reply:
Thanks for this tip @jaclync. It seems that `commitAndDismiss` is also called when the user presses the done button and there are no changes, so I just modified the `commitAndDismiss` method that now accepts a value `hasChangedData` c49dd1de443fa960f28be4df2a0ccfe1460e8c57 This fixed the problem 😸